torch/csrc/jit/passes/alias_analysis.cpp

```diff
@@ -496,6 +513,66 @@ void AliasDb::analyzeChunk(Node* node) {
   }
 }
 
+void AliasDb::analyzeWait(Node* node) {
+  // Do nothing, this node should have already been analyzed by the
+  // corresponding analyzeFork(). See that function for details.
+}
+
+void AliasDb::analyzeFork(Node* node) {
+  // We treat forks as if their subgraph ran immediately. Since the alias
+  // analysis is flow-insensitive, we can just assume that it runs here.
+  const auto subgraph = node->g(attr::Subgraph).get();
+  subgraphToOwner_.insert({subgraph, node});
+
+  const auto subgraphBlock = subgraph->block();
+  mapAliases(subgraphBlock->inputs(), node->inputs());
+  analyze(subgraphBlock);
+
+  // Propagate aliasing and write information from the subgraph outputs to the
+  // outputs of the corresponding aten::wait() calls, since that's where the
+  // values will eventually emerge.
+  const auto fut = node->output();
+  const auto subgraphWrites = getWrites(subgraph->block());
+  for (const auto& use : fut->uses()) {
+    const auto wait = use.user;
+    AT_ASSERT(wait->outputs().size() == subgraph->outputs().size());
+
+    // Propagate aliasing info.
+    mapAliases(wait->outputs(), subgraph->outputs());
+
+    // Propagate write information to the `wait` node.
+    //
+    // We need to do this for all writes in the entire subgraph, so that we
+    // disallow reorders past a call to "aten::wait".
+    //
+    // Consider the following Fork where the subgraph writes to %a:
+    //
+    //   %c : Future[Tensor] = prim::Fork(%a, %b) <-- writes to %a
+    //   ...
+    //   aten::wait(%c)
+    //   aten::use(%a)   <-- we can't move this node before the `wait` safely!
+    //
+    // Say we define the "live interval" of a fork the interval between the
+    // `fork` and its first corresponding `wait` (inclusive).
+    //
+    // Any writes in the subgraph can happen at any point in the live interval,
+    // so it's not safe to re-order any reads to those memory locations from
+    // outside the live interval to inside.
+    //
+    // In reality, any reads *inside* the live interval are undefined behavior,
+    // since the writes may or may not have been executed yet. But we'll let
+    // users do that and shoot themselves in the foot for now.
```

Comment:
Do we want to print a warning in this situation?

Reply:
I'm reluctant to incorporate alias analysis results in user-visible error reporting. We can't expect users to know about alias info/the IR/how we've mutated the graph up till now, so an error msg would be super confusing and probably not actionable.